### ISSUE TITLE
Support literal_unroll without type unification on loop body variables

### ISSUE THREAD

**Issue by kozlov-alexey** [numba/numba#6548]
- [] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Hi, 

I came across one use-case where literal_unroll currently fails, due to type-inference trying to unify types of the same variable from different iterations. This makes it not possible to support heterogeneous tuple-transformation functions, when you need to build a new tuple from existing one (e.g. by dropping or applying some function to it's elements ).

So this script:
```
import numpy as np
import numba
from numba.misc.special import literal_unroll

@numba.njit
def test_func_1():
    data = ('a', [2, 3], np.nan, 600)
    res = ()

    res = res + (data[0], )
    res = res + (data[1], )
    res = res + (data[2], )
    res = res + (data[3], )

    return res

@numba.njit
def test_func_2():
    data = ('a', [2, 3], np.nan, 600)
    res = ()

    for a in literal_unroll(data):
        res = res + (a, )

    return res

# res = test_func_1()
res = test_func_2()
print(f"DEBUG: res={res}")
```

fails with:
> numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
> Cannot unify Tuple() and UniTuple(unicode_type x 1) for 'res.5', defined at test_literal_unroll.py (25)
> 
> File "test_literal_unroll.py", line 25:
> def test_func_2():
>     <source elided>
> 
>     return res
>     ^
 
At the same type manually unrolling works fine, so `test_func_1` compiles and runs.
I think this is worth supporting, since using `tuple_setitem` mentioned in #5332 is not always enough and making unroll behave more like manually unrolled loop is worth trying. 

Also it seems below fails due to the same reason:

```
@numba.njit
def test_func_3():
    data = ('a', 'b', 1)

    for a in literal_unroll(data):
        new_a = (a, )
    return new_a
```

Best Regards,
Alexey.

**Comment by sklam** [comment]
This has to do with the a SSA-phi node that is merging multiple possible types for `res`:

```
res.4 = phi(incoming_values=[Var(res, iss6548.py:20), Var(res.5, iss6548.py:23)]
```

Initially an empty tuple from line `res = ()` and the fully extended result at the end.

The IR after the SSA pass looks like this:

<img width="1038" alt="Screen Shot 2020-12-04 at 2 16 51 PM" src="https://user-images.githubusercontent.com/1929845/101210607-62978680-363b-11eb-8a4a-2600193d62cc.png">

For this to work, the generated code cannot be a loop but a straightline code like the `test_func_1`.

**Comment by stuartarchibald** [comment]
I'm not sure that this is a bug, the feature is working as intended?

**Comment by kozlov-alexey** [comment]
@stuartarchibald Yes, I didn't think of it as a bug too, since I guess it was design decision to keep the phi node in the modified IR. But the above use-case is kind of corner-stone for applying to SDC use-cases (and as far as I understand there's no other way to support something like map function for tuples), so we would like you to consider this as a feature request.
